Bad build ids api changes

dnr opened the pull request:
**What changed?**
- Add `MarkBadBuild` option to `UpdateWorkerBuildIdCompatibilityRequest` and related messages
- Add build-id-based option to reset, including `RESET_TYPE_BUILD_ID` to `ResetType` in `BatchOperationReset`

**Why?**
Help recover from a bad deployment situation

**Breaking changes**
no

dnr committed:
add MarkBadBuild option

dnr committed:
add reset to build id options

bergundy commented on temporal/api/batch/v1/message.proto:
```diff
@@ -93,9 +93,13 @@ message BatchOperationDeletion {
+
+  // For reset_type == RESET_TYPE_BUILD_ID only:
+  // Resets to before the first workflow task processed by the specified build id.
+  string reset_build_id = 4;
```
Might be worth wrapping in a message in case we want to allow more options here. I'm already thinking that some users may not want to reset past the current run ID.

Sushisource replied:
+1

dnr replied:
good idea

dnr replied:
I made a new message for overall ResetOptions and put that flag (only current run) at the top level, since it could apply to other reset types. So then build_id didn't need a message again. I'm not opposed to a message there but are there really reset options that apply to build ids only?

bergundy commented on temporal/api/errordetails/v1/message.proto:
```diff
@@ -110,4 +110,6 @@ message WorkflowNotReadyFailure {
 message NewerBuildExistsFailure {
     // The current default compatible build ID which will receive tasks
     string default_build_id = 1;
+    // If this failure is due to the polling build id being marked as a bad build.
+    bool is_bad_build = 2;
```
Maybe change to a reason enum to support other reasons in the future?

dnr replied:
can you imagine anything else (that would fit generally under `NewerBuildExistsFailure` instead of some other error type)?

bergundy replied:
Not at the moment, but we didn't see this one coming either.
As a general guideline, I prefer enums even for dual states, you can't predict how this evolve so early on.

dnr replied:
Hmm. Except without another example it's not clear if these should be exclusive (like an enum) or orthogonal (like other fields). E.g. a build can be both not-default and also a bad build.

dnr replied:
I'm still not sure about this one. Note that these two fields are not exclusive and actually maybe we wouldn't set `default_build_id` if the default was marked bad without a replacement. I.e. the options are:
- there is a new default build id, but the one you polled with is not marked bad
- there is a new default build id, also btw the one you polled with is bad
- there is no default since there was one and it got marked bad, so now nobody gets any tasks

At this point it's really the name of the message/error type that's not accurate, but I don't want to change that now

dnr replied:
We already have the custom error type, I'm just adding a field to it

bergundy replied:
I know, this is more of a lesson for next time.

bergundy commented on temporal/api/workflowservice/v1/request_response.proto:
```diff
@@ -663,12 +663,15 @@ message ResetWorkflowExecutionRequest {
+        // The id of a `WORKFLOW_TASK_COMPLETED`,`WORKFLOW_TASK_TIMED_OUT`, `WORKFLOW_TASK_FAILED`, or
+        // `WORKFLOW_TASK_STARTED` event to reset to.
+        int64 workflow_task_finish_event_id = 4;
+        string reset_build_id = 7;
```
same, consider making this a message instead of a string.

bergundy commented on temporal/api/taskqueue/v1/message.proto:
```diff
@@ -95,6 +95,10 @@ message StickyExecutionAttributes {
+    // Build ids that have been marked as bad. If the last element in build_ids is also in
+    // bad_build_ids, then no tasks will be dispatched on this version set until the default is
+    // changed.
+    repeated string bad_build_ids = 2;
```
In the internal representation, you can add a new state in the enum, this is fine here though.

dnr replied:
yup, that was the plan. we could try to transition the string list to a message but... seems more trouble than it's worth. I don't like the repetition but it's the simplest thing here. I guess it could be a parallel array of bools but that's worse

bergundy replied:
Yup, this is the best option given the current API, let's keep it and avoid breaking existing clients.

dnr commented on temporal/api/batch/v1/message.proto:
```diff
@@ -93,9 +93,13 @@ message BatchOperationDeletion {
-}
\ No newline at end of file
+
+  // For reset_type == RESET_TYPE_BUILD_ID only:
```
I wish this was a oneof in the first place :(

dnr replied:
I just went and made it a oneof. I think it's the cleanest way to deal with the new reset options.

dnr committed:
BuildIdResetOptions

yiminc commented on temporal/api/workflowservice/v1/request_response.proto:
```diff
@@ -663,12 +663,16 @@ message ResetWorkflowExecutionRequest {
+        // `WORKFLOW_TASK_STARTED` event to reset to.
+        int64 workflow_task_finish_event_id = 4;
+        // Reset to before a specific build id.
+        temporal.api.common.v1.BuildIdResetOptions reset_build_id = 7;
```
Currently, we have reset option to reset to first or last workflow task. And the caller (batch workflow) has to figure out which workflow task event_id that option maps to. That is different from this new option which history service has to figure out which event_id it maps to.
I like this new approach, but it make it inconsistent with existing ones. Probably should (in future) add the first / last workflow task as option here and move those logic from batch workflow to history service.

yiminc replied:
Just realize this is already requested as https://github.com/temporalio/temporal/issues/1338

dnr replied:
I see. Do you think we should extract all this into a whole new message and migrate to that here? Otherwise we have a oneof here, and then an enum for the batch request forever... instead we could make a new message and use it for both, and deprecate the old fields.

Or we could have the batch workflow look up the reset point, but that seems silly since history owns it.

dnr replied:
I just made a new message and used it in both places. It'll require some compatibility code to translate but it shouldn't be too bad.

dnr committed:
add common ResetOptions

dnr committed:
update comments

dnr committed:
use ResetOptions only for batch for now, not ResetWorkflowExecutionRequest

dnr committed:
add build id to ResetPointInfo

dnr commented:
I changed this to use the new ResetOptions only for the batch request, not the individual request. Using it for the individual request could come later, although maybe there are good reasons to have them all go through batch (so that we do the digging through history on the worker instead of frontend or history).

I also added a build id to ResetPointInfo so it has both a build id and binary checksum now, and we can remove binary checksum later.

There are a few situations that may require some more changes here, but this should be good to go for phase 1.

Sushisource commented on temporal/api/batch/v1/message.proto:
```diff
@@ -92,10 +92,16 @@ message BatchOperationDeletion {
-}
\ No newline at end of file
+
+  // New API (exclusive with `reset_type` and `reset_reapply_type`):
```
IMO this comment should just say what this is for, and the comments for the older things can call out that they are old/deprecated.

Maybe keep the exclusivity mention, but don't need to say it's "New" I think.

Sushisource commented on temporal/api/common/v1/message.proto:
```diff
@@ -149,3 +151,29 @@ message WorkerVersionCapabilities {
+message ResetOptions {
+    // Which workflow task to reset to.
+    oneof target {
+        // Resets to event of the first workflow task completed, or if it does not exist, the event after task scheduled.
```
What if there's no event after task scheduled either?

I think this wording is slightly more clear:
```suggestion
        // Resets to the first `WORKFLOW_TASK_COMPLETED` event, or if it does not exist, the first `WORKFLOW_TASK_SCHEDULED` event.
```

dnr replied:
I'll just say "first workflow task completed or started event", I think that's enough for this purpose.

Sushisource commented on temporal/api/common/v1/message.proto:
```diff
@@ -149,3 +151,29 @@ message WorkerVersionCapabilities {
+    oneof target {
+        // Resets to event of the first workflow task completed, or if it does not exist, the event after task scheduled.
+        google.protobuf.Empty first_workflow_task = 1;
+        // Resets to event of the last workflow task completed, or if it does not exist, the event after task scheduled.
```
Same comment as w/ first

Sushisource commented on temporal/api/common/v1/message.proto:
```diff
@@ -149,3 +151,29 @@ message WorkerVersionCapabilities {
+        // The id of a specific `WORKFLOW_TASK_COMPLETED`,`WORKFLOW_TASK_TIMED_OUT`, `WORKFLOW_TASK_FAILED`, or
+        // `WORKFLOW_TASK_STARTED` event to reset to.
+        int64 workflow_task_id = 3;
+        // Resets to before the first workflow task processed by this build id.
```
Which event, specifically, before the first WFT processed by the build ID?

dnr replied:
As I understand it, it doesn't matter whether it's the scheduled, started, or completed/failed/timedout event, the effect is the same for all. I'll just say "first workflow task processed by this build id"

Sushisource commented on temporal/api/workflowservice/v1/request_response.proto:
```diff
@@ -1110,10 +1110,26 @@ message UpdateWorkerBuildIdCompatibilityRequest {
+    message MarkBadBuild {
+        // The build id to mark as bad.
+        string bad_build_id = 1;
+        // If promote_build_id is set, then also promote it to be the default of the set.
+        // If it's not set, then no tasks will be dispatched on the task queue set until another
+        // build id is made the default by another call to UpdateWorkerBuildIdCompatibility that
+        // sets/adds a new default for the set.
```
This comment implies that `bad_build_id` is the current default for it's set, but, that might not be the case. Presumably you really only care to do this if it is, but, it's worth mentioning that.

Sushisource commented on temporal/api/workflowservice/v1/request_response.proto:
```diff
@@ -1142,6 +1158,14 @@ message UpdateWorkerBuildIdCompatibilityRequest {
+        // Mark a build as "bad" for a particular task queue. This means that:
+        // 1. the build will not receive any more tasks (workflow or activity)
+        // 2. it may not be promoted to the set default anymore
+        // See the MarkBadBuild for more information.
```
```suggestion
        // See MarkBadBuild for more information.
```

Sushisource reviewed (commented):
Structurally makes sense to me. Some stuff to fix in the docstrings.

dnr reviewed (commented):
Splitting this into a separate PR for reset, and the rest can come later

dnr commented:
See #333
